[COT/HPOS] Fix trashing/untrashing of orders

Konamiman opened the pull request:
-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This pull request fixes some small bugs:

1. Trashing an order when COT was authoritative and sync was enabled didn't update the update date of the COT record. This caused that the next time the order was read it was untrashed (per the change introduced in https://github.com/woocommerce/woocommerce/pull/34465).
2. Similarly, trashing an order when COT was authoritative and sync was enabled wasn't trashing the equivalent entry in the posts table.
3. When an order was untrashed the corresponding `_wp_trash_meta_status` and `_wp_trash_meta_time` meta records weren't deleted. Thus if the order was trashed and untrashed repeatedly new meta rows would be created every time that would end up piling up in the meta table.
4. The change introduced in https://github.com/woocommerce/woocommerce/pull/34465 was unnecessarily re-syncing the orders when the update dates in COT and in the posts table were identical.
5. It was possible to load the order details in Admin when the order was trashed via direct URL. Now this shows an error in the same way as when posts is authoritative.

Closes #35089.

### How to test the changes in this Pull Request:

1. With COT enabled and sync active trash an order.
2. Verify that you can't access the order via Admin (navigate to `/wp-admin/admin.php?page=wc-orders&id=<order id>&action=edit` and you'll get an error).
3. Verify that the corresponding entry in the posts table is also trashed.
4. Go to the trashed orders view, then to the all orders view, verify that the order doesn't get untrashed automatically.
5. Untrash the order and verify that the order has the proper status and that there aren't `_wp_trash_meta_status` entries in the `wp_wc_orders_meta` table for the order.
6. Verify that the sync in the opposite direction (posts to COT) still works for the trash-untrash cycle (note that in this case you need to manually trigger the sync).
7. Manually alter the entry in posts table for an order, load the order, and verify that the data in COT has been synced with the changes you introduced (aka verify that https://github.com/woocommerce/woocommerce/pull/34465 still works).

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Fix trashing/untrashing orders when using COT:

- Trashing wasn't changing the update date on the orders table record.
- Trashing wasn't applying trash on the corresponding post.
- Untrashing wasn't actually deleting _wp_trash_meta_status in metadata.

Konamiman committed:
Fix: maybe_sync_order was syncing even with identical update dates.

Konamiman committed:
PHPCS is happy now.

Konamiman committed:
WP_DIE when attempting to load a trashed order in Admin
(same behavior as when using the posts table)

Konamiman committed:
Add changelog file

barryhughes commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -1059,9 +1059,7 @@ private function maybe_sync_order( \WC_Abstract_Order &$order, \WC_Abstract_Orde
-		if ( $order_modified_date > $post_order_modified_date ) {
-			return;
-		} else {
+		if ( $post_order_modified_date > $order_modified_date ) {
```
If we're just simplifying the code down to keep things nice and clean, this would maintain the previous behaviour (in which we would also call `migrate_post_record` when the modification dates are equal):

```suggestion
		if ( $post_order_modified_date >= $order_modified_date ) {
```

This also fixes the three `OrdersTableDataStoreTests` fails.

Konamiman replied:
I neglected the tests, my fault. But the removal of the `=` comparison is on purpose (the "cleanification" of the code was just a side effect): if both records have the exact same update date it means that they are in sync and thus there is no need to migrate, right?

barryhughes replied:
Due to the check at the top of the method [(is the post different from the order?)](https://github.com/woocommerce/woocommerce/pull/35125/files#diff-8c34c1a6f743002d1f2db13c299d2086d0209889f72abc86bef3bd734abb2585R1044-R1046) I assumed that if we reach this point then there *is* some kind of difference between the two records that ought to be reconciled?

Konamiman replied:
After discussing with @vedanshujain I'm applying this change. Turns out that data comparison is needed too when the dates are equal, since data in the metadata table could have been modified as well.

barryhughes commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -1757,13 +1755,21 @@ public function trash_order( $order ) {
+
+		$data_synchronizer = wc_get_container()->get( DataSynchronizer::class );
+		if ( $data_synchronizer->data_sync_is_enabled() ) {
+			wp_trash_post( $order->get_id() );
```
> Untrash the order and verify that the order has the proper status

☝🏽 This part is not working for me (COT authoritative, sync enabled).

It looks to me like we store the pre-trash status and trash time as order meta data, just [a few lines above](https://github.com/woocommerce/woocommerce/pull/35125/files#diff-8c34c1a6f743002d1f2db13c299d2086d0209889f72abc86bef3bd734abb2585R1741-R1754) this point. Then execution reaches here and we trash the corresponding post record ... the problem is that those bits of meta data never makes it across to the post meta table. The result is the order is not restored to its original status when it is untrashed.

Steps to replicate:

1. Check out this branch, set COT to authoritative, enable sync.
2. Select an order that is **not** *pending payment,* note its status and trash it.
3. Visit the trashed orders view.
4. Untrash the order, note it will have been incorrectly restored to *pending payment* status (which is the default we use when the original status is not known).

Konamiman replied:
I can't reproduce this. I have tried with "on hold", "processing" and "complete" and in all cases:

1. Trashing the order generates the proper meta items in both meta tables:

![image](https://user-images.githubusercontent.com/937723/196370650-4fa12172-1132-4c9c-a70b-056156bf109f.png)

2. Untrashing the order restores it to the proper state.

What I have noticed is that after untrashing the entry in `wp_postmeta` is still there. Maybe because it doesn't have the `wc-` prefix in the status name?

barryhughes replied:
Interesting ... I wonder why the difference. Let me dig just a little bit further and/or verify using a fresh/'blank' environment. Perhaps something particular to me environment is causing an issue.

barryhughes replied:
I still see this issue. I'll send you some details to help you see it.

Konamiman replied:
Fixed. Explanation of what was going on in [the commit message](https://github.com/woocommerce/woocommerce/pull/35125/commits/f538c7c272fa5c02010aea62a281462cf5b8dccb).

btw I wasn't able to reproduce initially because I was testing with an existing order that already had dummy meta status in the db, as soon as I tried with a brand new order I was able to see the the issue.

barryhughes replied:
Yep, works :+1:

Konamiman committed:
Revert change in order records date comparison

Data comparison is needed too when the dates are equal, since data in the metadata table could have been modified as well.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

Konamiman committed:
Fix handling of order status and trash meta when untrashing.

Untrashing an order was setting it to the default status of "pending"
instead of the previous status. This is what was happening:

1. Order is trashed, order meta table gets the value (e.g.) "on-hold"
   for the _wp_trash_meta_status key
2. Post is trashed, post meta table gets "wc-on-hold"
3. Order is re-read, but then so is posts, and the order meta value is
   updated to "wc-on-hold"
4. On untrash the previous order status from order meta table is
   prefixed with "wc-", thus it becomes "wc-wc-on-hold", which is not
   a valid status
5. Not having a valid status, the order is restored as the default
   status of "pending payment"

To fix this, the previous status is stored in order meta table
with "wc-", and it doesn't get prefixed on untrash.

Also, wp_untrash_post is called on order untrash, otherwise the
trash metadata keeps dangling and being synced back and forth.

barryhughes commented:
Conflict to resolve then we can merge :+1:

Konamiman committed:
Merge branch 'trunk' into fix/35089

Konamiman committed:
Add unit tetst for the trash-untrash cycle
